Static color sliders / field

ericmehl opened the pull request:
This adds the option to set the color sliders and field to dynamically update. In dynamic mode, which is how they have been thus far, the sliders show the color you will get if you move the indicator to the position at that color. They update as you change other channels.

The default mode is now non-dynamic.

With dynamic mode turned off, the slider shows the pure 0-1 (-1 to 1 for temperature and magenta) range for that channel only. So the red slider will always go from black to fully red, and similar for the other colors. The only exception is the saturation slider and the color field when the static component is set to hue. These will still update because they depend on the hue to show their full range.

With dynamic mode off, we can also optimize when we calculate the gradients and field so they don' t update as often. Not updating the gradient gives about a 90% speedup which does feel a little snappier when dragging the sliders.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

murraystevenson commented on python/GafferUI/ColorChooser.py:
```diff
@@ -153,42 +160,44 @@ def getDynamicColors( self ) :
 
-		displayTransform = self.displayTransform()
+		if self.__gradientToDraw is None :
+			self.__gradientToDraw = QtGui.QLinearGradient( 0, 0, size.x, 0 )
```
With the optimization we run into a little trouble if the slider is resized. The gradient doesn't update to account for the new width.

![image](https://github.com/user-attachments/assets/7db4bfa6-0eab-4a7a-968c-54856641ed0b)

ericmehl replied:
Right, good catch. I don't see a resize signal on `Widget` to use, so in https://github.com/GafferHQ/gaffer/pull/6111/commits/dd5aad0e71213a37c7cf40b0f0f69616a40999be I just track the size and redraw when it has changed.

ericmehl commented:
Based on offline conversations, I've pushed a few new commits today. I changed the name of the variables to be variations on `dynamicBackground` instead of `dynamicColor` to clarify that it's the background only, not the actual color being chosen that is dynamic.

I also changed the wording in the settings menu to be "Sliders / Dynamic Backgrounds" for the same reason.

And lastly, I changed the default style to be dynamic instead of static. That's based on some user polling that favored the dynamic style.

Ready for a new look!

murraystevenson reviewed (changes requested):
Thanks Eric! A couple of nitpicky thoughts inline, but otherwise this is looking pretty good to me.

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -845,6 +898,8 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
 		self.__updateUIFromColor()
 		self.__updateComponentIcons()
 
+		self.setDynamicSliderBackgrounds( True )
```
Yeah, that's nice to have removed and the defaults clarified. Done in https://github.com/GafferHQ/gaffer/pull/6111/commits/64a5250fe8f5f32dbcfea5aa208ccbf0d74eb369 and https://github.com/GafferHQ/gaffer/pull/6111/commits/56bff8e5ec90e42dae89864effcc7fb1c09099f4

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -1018,6 +1086,16 @@ def __optionsMenuDefinition( self ) :
 				}
 			)
 
+		result.append( "/__sliders__", { "divider": True, "label": "Sliders" } )
```
For this user-facing part, we talked on a call (apologies I forgot you weren't on that one) about just using "Sliders" as a compromise between having a long but accurate description and being succinct. A slight additional justification is that you could see the color field as being a 2d slider, which is maybe not stretching things too much?

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -920,6 +975,18 @@ def getColorFieldVisible( self ) :
 
 		return self.__colorField.getVisible()
 
+	def setDynamicSliderBackgrounds( self, dynamic ) :
```
I forgot you weren't on the call where we talked about this, or at least the user facing part (the "Sliders" label mentioned below). I took the rationale below, especially about the flexible use of "slider" to also avoid making the API portion unwieldy.

For the sake of discussion, maybe `setDynamicWidgetBackgrounds()`? Or `setWidgetBackgroundsDynamic()`? `setSliderBackgroundsDynamic()`?

murraystevenson replied:
Cool, I appreciate the rationale and as we're going to keep "Sliders" in the UI, it makes sense to keep it consistent here.

ericmehl commented on python/GafferUI/ColorChooser.py:
```diff
@@ -1018,6 +1086,16 @@ def __optionsMenuDefinition( self ) :
+			"/Dynamic Backgrounds",
+			{
+				"command": Gaffer.WeakMethod( self.setDynamicSliderBackgrounds ),
+				"checkBox": self.getDynamicSliderBackgrounds(),
```
Done in https://github.com/GafferHQ/gaffer/pull/6111/commits/7e709dbef621d42068de7c156da11cb4126d746f. 

Though I'm not super happy with the cosmetic formatting of that solution. I couldn't find a way to get it formatted better / independently of the source line layout the way the plug tooltips do.

murraystevenson replied:
To match the more flexible formatting of the plug tooltips, we'd likely need a `GafferUI.DocumentationAlgo.markdownToHTML( action.statusTip() )` here :
https://github.com/GafferHQ/gaffer/blob/05aef9e3c6b425dd0de40409c4fe047709895569/python/GafferUI/Menu.py#L729
But that could start doing undesirable things in tooltips elsewhere such as wrapping file paths and the like, so I'd avoid that for now.

If you're unhappy with the current text, we could simplify this down to something more along the lines of :

"Dynamic backgrounds show the color that will result from moving the indicator to a given position."

Feel free to make a call on that while you squash everything down ready for merging. Thanks!

ericmehl replied:
I just went with the original text I had, I think it's a better explanation and not worth worrying over the formatting.

ericmehl commented:
Pushed a few new commits to address the comments above. I think except for the API naming we should be in good shape.

ericmehl commented on Changes.md:
```diff
@@ -5,6 +5,7 @@ Improvements
 ------------
 
 - Instancer : Added support for 64 bit ints for ids ( matching what is loaded from USD ).
+- ColorChooser : Added an option to enable or disable dynamic colors updates of the color sliders and color field. If enabled, the widget colors update to show the color that will result from moving the indicator to a given position. If disabled (the default), the widget always shows the range of values for the channel(s) alone.
```
Done and squashed into the original commits : https://github.com/GafferHQ/gaffer/pull/6111/commits/2dd9e1806f0567774af2f879a5546751a4bd0248 and https://github.com/GafferHQ/gaffer/pull/6111/commits/d35108ebbb0ba399a4853927f2032b6821d1d3fb

ericmehl commented:
Thanks Murray! Squashed down and verified everything matches up to the pre-squash state. Merging
